src/Doctrine/Orm/State/LinksHandlerTrait.php

```diff
@@ -40,7 +40,7 @@ private function handleLinks(QueryBuilder $queryBuilder, array $identifiers, Que
         $doctrineClassMetadata = $manager->getClassMetadata($entityClass);
         $alias = $queryBuilder->getRootAliases()[0];
 
-        $links = $this->getLinks($entityClass, $operation, $context);
+        $links = $this->getLinks($context['resource_class'] ?? $entityClass, $operation, $context);
```

Comment:
From what I understand the problem occurs only when it's an embed resource? I'm wondering though why this `entityClass` is wrong, are you able to trace that call and fix it in the class responsible from calling this? Basically at https://github.com/api-platform/core/blob/3.4/src/Doctrine/Orm/State/ItemProvider.php#L74 it should already be the correct class.